Allows encoding while constructing HTTP request for sending notification

mihirsoni opened the pull request:
*Issue #, if available:*
#33 
*Description of changes:*

We were not encoding the message while publishing, using the [StringEntity](https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/entity/StringEntity.html#StringEntity(java.lang.String,%20org.apache.http.entity.ContentType)) constructor with the mime-type and charset.

* Chime / Slack Defaults `application/json`
* Custom Webhook , Default to `application/json` , user can choose other values.

Sample Chime notification : 
![Screen Shot 2019-04-24 at 9 45 55 AM](https://user-images.githubusercontent.com/578861/56677659-d6a73580-6675-11e9-9f71-4e45f882f625.png)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

vamshin commented on notification/src/main/java/com/amazon/opendistroforelasticsearch/alerting/destination/client/DestinationHttpClient.java:
```diff
@@ -52,7 +53,8 @@
     private static final int TIMEOUT_MILLISECONDS = (int) TimeValue.timeValueSeconds(5).millis();
     private static final int SOCKET_TIMEOUT_MILLISECONDS = (int)TimeValue.timeValueSeconds(50).millis();
-
+    private static final String DEFAULT_CONTENT_TYPE_KEY = "Content-Type";
```
variable name could be just **CONTENT_TYPE_KEY**

vamshin commented on notification/src/main/java/com/amazon/opendistroforelasticsearch/alerting/destination/client/DestinationHttpClient.java:
```diff
@@ -90,6 +92,7 @@ public String execute(BaseMessage message) throws Exception {
     private CloseableHttpResponse getHttpResponse(BaseMessage message) throws Exception {
         URI uri = null;
         HttpPost httpPostRequest = new HttpPost();
+        String entityContentType = DEFAULT_CONTENT_TYPE_VALUE;
```
entityContentTypeValue?

vamshin commented on notification/src/main/java/com/amazon/opendistroforelasticsearch/alerting/destination/client/DestinationHttpClient.java:
```diff
@@ -98,20 +101,24 @@ private CloseableHttpResponse getHttpResponse(BaseMessage message) throws Except
             if(headerParams == null || headerParams.isEmpty()) {
-                // set default header
-                httpPostRequest.setHeader("Content-Type", "application/json");
+                // set default header and content Type
```
content-type is part of header. Lets keep the comment as it is?

vamshin commented on notification/src/test/java/com/amazon/opendistroforelasticsearch/alerting/destination/CustomWebhookMessageTest.java:
```diff
@@ -79,6 +82,49 @@ public void testCustomWebhookMessage() throws Exception {
+        DestinationFactoryProvider.setFactory(DestinationType.CUSTOMWEBHOOK, customDestinationFactory);
+
+        Map<String, String> headerParams = new HashMap<String, String>();
+        headerParams .put("Content-Type", "application/xml");
```
np: spaces headerParam**s .p**ut

vamshin commented on notification/src/test/java/com/amazon/opendistroforelasticsearch/alerting/destination/CustomWebhookMessageTest.java:
```diff
@@ -79,6 +82,49 @@ public void testCustomWebhookMessage() throws Exception {
 
+    @Test
+    public void testCustomWebhookMessageWithHeaders() throws Exception {
+
```
It would be good to assert we are passing the expected header in the HttpPostRequest.

mihirsoni commented:
@vamshin I just had to use charset while in entity as we've already set in headers on the POST so it wasn't required. I've removed other changes and added the charset for Body String.

Pupkur commented:
Can you please explain how to set up character encoding?
Add header to "Header information"?
key - charset
value - UTF-8

And which version will fix this issue?

mihirsoni commented:
> Can you please explain how to set up character encoding?

Generally , you can setup like this 

`Content-Type: text/html; charset=utf-8`

> And which version will fix this issue?

This will be released in next Opendistro version.
